### DISCUSSION TITLE
Request for Assistance: Localization of Kura Web Interface to Chinese

### DISCUSSION THREAD

**Discussion by ik98992** [eclipse-kura/kura#5430] [Q&A]
Hi @mattdibi @MMaiero ,

 I am currently utilizing the Eclipse Kura platform and have noticed that the web interface is predominantly displayed in English. I am interested in localizing it to Chinese to better cater to local users.
﻿
I have attempted various methods, including custom translation files and browser plugins, but have encountered difficulties. Therefore, I am reaching out to the community for assistance and guidance on how to effectively localize the Kura web interface to Chinese.
﻿
I am seeking advice on best practices or guidelines for translating the Kura web interface. Are there any tools or plugins developed by the community that could assist in achieving this goal? Additionally, I am curious to know if anyone within the community has undertaken similar localization efforts and could share their experiences and recommendations.

Any feedback is welcomed. Many thanks in advance!

**Comment by mattdibi** [comment]
Hello there @ik98992, thank you for reaching out!

The current best practice to add support for a new language in Kura is to add the relevant translation in the [messages properties](https://github.com/eclipse-kura/kura/tree/develop/kura/org.eclipse.kura.web2/src/main/resources/org/eclipse/kura/web/client/messages). You can look at the [`Messages_ja.properties` for reference](https://github.com/eclipse-kura/kura/blob/develop/kura/org.eclipse.kura.web2/src/main/resources/org/eclipse/kura/web/client/messages/Messages_ja.properties). Once added, you'll need to rebuild.

Keep in mind that Metatypes won't be translated. This is a limitation of the current WebUI.

The WebUI will then show the language depending on your system settings. For testing purposes, in the past, we used a couple of different locale switchers browser extensions (see for example [this one](https://chromewebstore.google.com/detail/locale-switcher/kngfjpghaokedippaapkfihdlmmlafcc?hl=en-GB&utm_source=ext_sidebar))

Hope this helps!

**Reply by MMaiero** [reply]
To add on top of what @mattdibi said, we are willing to expand the locales of our project and receiving a contribution for Chinese locale would be huge for us and greatly appreciated.
We are just few weeks before a new release, and this could be a great contribution to add to our upcoming Eclipse Kura 5.6.0 release.

**Reply by mattdibi** [reply]
Hi @ik98992, 

FYI @spnettec maintains a fork with Chinese localisation. Take a look here: https://github.com/spnettec/kura/tree/heyoulin-kura/kura/org.eclipse.kura.web2/src/main/resources/org/eclipse/kura/web/client/messages

@spnettec in case you wanted to contribute the Chinese localisation to Kura it would be greatly appreciated!

**Reply by spnettec** [reply]
@MMaiero @MMaiero 
It's my honor to contribute my code.
I have submitted my code in past two years ago. But the code modification is very large. The community did not approve my changes.
I need to separate the localization part from my branch. Since I am very busy recently it may take a long time. 
I can take the time to submit the localization core code. 
The translation part is very laborious. Each individual module needs to be translated. Since my branch is quite different from the community branch i cannot submit individual module translation files.

**Reply by mattdibi** [reply]
First and foremost, @spnettec thank you for the feedback and the commitment. This is greatly appreciated 🙏 

🤔 I think, as an initial contribution, the UI part (i.e. the .properties file I linked in my messages above) would be a great starting point.

Regarding the individual modules (I think you're talking about each individual metatype for the services scattered around the codebase... feel free to correct me if I'm wrong) we might plan a separate way. I'm not knowledgeable on the metter and therefore I would ask @nicolatimeus and @MMaiero for ideas. Do we need something, code side, to better handle this type of translations/localisation?

**Comment by spnettec** [comment]
This has nothing to do with the .properties file which can be generated using translation tools. In fact we need to expand the localization function. Some locale strings is in the hard code. If not expand the locale function  we can not do this
<img width="1693" alt="image" src="https://github.com/user-attachments/assets/6241f1bd-8067-4be8-bab7-94e276b27e12">
<img width="1749" alt="image" src="https://github.com/user-attachments/assets/4258a946-7596-4183-907b-03b478f9b157">

**Reply by spnettec** [reply]
We need to refactor ComponentConfiguration class. This would affect the reconstruction of the web ui module. This is a big change

**Reply by mattdibi** [reply]
Hi there @spnettec, sorry for taking so long to answer.

We do agree with the localisation issues and the fact that we need to expand this functionality. Some initial effort was done [here](https://github.com/eclipse-kura/kura/tree/develop/kura/org.eclipse.kura.localization) and [here](https://github.com/eclipse-kura/kura/tree/develop/kura/org.eclipse.kura.localization.resources) but we never had the time to commit to a complete update for all our modules.

Can you share some details on how did you achieve the localisation on your fork? Do you think it would be possible to plan a way to split it in manageable chunks and contribute it to the main Kura repo (if you think it has value and you're willing to undertake this effort of course)?

Feel free to come by our [Kura dev meeting](https://github.com/eclipse-kura/kura/discussions/4841) in case you want to discuss this in real time.
